### PR TITLE
CP-49029: Instrument `xapi_session.ml`  with tracing

### PR DESCRIPTION
**Pull request by GabrielBuica**
Instruments `xapi_session.ml` to create span around session logins.

Adds new attribute for spans:

- `xs.xapi.session.originator`

where available.

### CODE REVIEW

File: ocaml/xapi/xapi_session.ml [hunk 1]
```diff
@@ -303,8 +303,9 @@ let do_local_change_password uname newpwd =
 let trackid session_id = Context.trackid_of_session (Some session_id)
 
 (* finds the intersection between group_membership_closure and pool's table of subject_ids *)
-let get_intersection ~__context subject_ids_in_db subject_identifier
-    group_membership_closure =
+let get_intersection ~(__context : Context.t) subject_ids_in_db
```
Comment (psafont) [inline ocaml/xapi/xapi_session.ml:306]: Why add the type here?

**Review by lindig — APPROVED** [review]

**Comment by psafont** [comment]
What are the consecuences of missing the originator? If it's of little inconsequence it might be fine to have the optional argument, if it's important to not miss it, it probably needs to be forced. Independently I'd prefer to see an originator argument for with_tracing, and doing the covnersion there between originator and attributes:
```ocaml
let with_tracing ~context ~originator name f =
  let open Tracing in
  let parent = context.tracing in
  let span_attributes = Tracing.attributes_of_originator originator in
  match
    start_tracing_helper ?span_attributes:attributes (fun _ -> parent) name
  with
  ...
```

File: ocaml/xapi/xapi_session.ml [hunk 2]
```diff
@@ -742,6 +758,10 @@ let slave_local_login_with_password ~__context ~uname ~pwd =
       against the local superuser credentials
 *)
 let login_with_password ~__context ~uname ~pwd ~version:_ ~originator =
+  Context.with_tracing
+    ~attributes:[("xs.xapi.session.originator", originator)]
```
Comment (mg12) [inline ocaml/xapi/xapi_session.ml:762]: this is repeated above, we need a single place that creates this attribute, perhaps a function attributes_of_originator in session.ml: Session.Tracing.attributes_of_originator?

File: ocaml/xapi/xapi_session.ml (hunk 1 above)
Comment (GabrielBuica) [reply ocaml/xapi/xapi_session.ml:306]: It's a leftover that I forgot to remove. Thanks!

File: ocaml/xapi/xapi_session.ml (hunk 2 above)
Comment (GabrielBuica) [reply ocaml/xapi/xapi_session.ml]: @psafont  @mg12 Something like this? https://github.com/xapi-project/xen-api/pull/5621/commits/1aedd6ddb9867debe7cbc3e5ae9e3e5c9ba1d17d

**Review by psafont — APPROVED** [review]

**Review by mg12 — APPROVED** [review]
lgtm: @GabrielBuica needs squashing before merge

**Pull request merged**